pg_shard.c

```diff
@@ -1107,6 +1105,17 @@ PgShardExecutorStart(QueryDesc *queryDesc, int eflags)
 			bool topLevel = true;
 			LOCKMODE lockMode = NoLock;
 			EState *executorState = NULL;
+			bool zeroShardQuery = list_length(distributedPlan->taskList) == 0;
+
+			/* if query involves zero shards, just let it hit local table */
+			if (zeroShardQuery)
+			{
+				Plan *originalPlan = distributedPlan->originalPlan;
+				plannedStatement->planTree = originalPlan;
+
+				PgShardExecutorStart(queryDesc, eflags);
```

Comment:
Sorry, the DRY principle has a pathological grip on me. I'll move the "decide which hook to run" into its own function, though… we'll have it in three places in this function otherwise (copy-pasting it again was why I went with the recursive call).

I think I can rework this the way you asked without adding too many lines of code.